### PR TITLE
allow Generator to be Send when sound

### LINKED ISSUE #13: Make boxed generators `Send` + `Sync`
(This may already be the case and I'm ignorant of how to make it work.)

I work on a IO library that uses generators to implement transactions. Its is used by different runtimes (Java, Swift, …) across [an FFI boundary that requires types to be `Send` + `Sync`](https://github.com/mozilla/uniffi-rs). Here's a simple example:

```rust
#[generator(yield(Vec<u8>), resume(Vec<u8>))]
fn version() -> Result<u16, TransactionError> {
    let apdu: apdu::Command = wca::WCA::Version.try_into()?;
    let response = yield_!(apdu.into());
    let value = apdu::Response::from(response)
        .payload
        .try_into()?;
    let result = Some(u16::from_be_bytes(value))
        .ok_or(TransactionError::InvalidResponse)?;

    Ok(result)
}
```

n.b. I did see genawaiter supports `Sync` + `Send` generators; but, it appears to be unmaintained.

### PR DESCRIPTION
**Pull request by pgavin**
I believe this fixes https://github.com/danielhenrymantilla/next-gen-rs/issues/13

### CODE REVIEW

**Comment by danielhenrymantilla** [comment]
Apologies for the slow answer; so the reason I did not do this is that auditing the code of `YieldSlot` in presence of multi-threading is quite more subtle! And I haven't taken the time to do it properly yet. Hence the conservative stance not to assume thread-safety, in case of a niche usage which could be problematic.

But I wouldn't be against the option for the downstream user themselves, to "remotely `unsafe`ly assert that this crate is indeed thread-safe", if _they_ are sure this is fine.

In other words, you could add a `"unsafe-assume-thread-safety"` Cargo feature on the crate, and then `#[cfg()]`-gate your impl with it (you could add a section to the main docs talking about it).

That way I'd be more welcome to merging this directly/quickly, while waiting for me to take the time to properly think about the repercussions in multi-threaded code. If we end up convinced enough that it is fine, we can later remove the `cfg`-gating on the `impl` 🙂
